File: TransForm_Kit/Quantization/quantization.py
# Copyright 2019 Inspur Corporation. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

import torch
import numpy as np
import struct
import math
from net_structure import structure_hook
from model_loader import load_model
import sys, os

def QuantizeLinear(x):
    Max = torch.max(abs(x))
    Limiter = pow(2,n-1) - 1
    scale = 1
    if Max != 0:
        scale = torch.round(Limiter/Max)
    return (scale)

def QuantizeForShift(x):
    Max = np.max(abs(x))
    Q = 0
    if Max > 0:
        Q = np.log2((127/Max))
        if Q<0:
            Q = np.floor(Q)
        else:
            Q = np.round(Q)
        out = x*pow(2,Q)
        while np.max(out)>127 or np.min(out)<-128:
            Q = Q-1
            out = x*pow(2,Q)
    return(Q)

def QuantizeChannel(style, x):
    shape = x.shape
    if style == 'shift':
        power = []
        if len(shape) == 1:
            for i in range(x.shape[0]):
                Q = QuantizeForShift(x[i])
                power.append(Q)
        else:
            for i in range(x.shape[1]):
                Q = QuantizeForShift(x[:,i])
                power.append(Q)
        power = np.array(power)
        mean = power.mean()
        for i in range(power.shape[0]):
            if power[i] > 0:
                if power[i] > mean:
                    power[i] = math.floor(mean)
            if power[i] < 0:
                if power[i] < mean:
                    power[i] = math.ceil(mean)
        return (power)
    if style == 'simple' or 'winograd':
        scale_list = []
        if len(shape) == 1:
            for i in range(x.shape[0]):
                scale = QuantizeLinear(x[i], n)
                scale_list.append(scale)
        else:
            for i in range(x.shape[1]):
                scale = QuantizeLinear(x[i], n)
                scale_list.append(scale)
        return (scale_list)

    

if __name__ == "__main__":
    
    batch_size = 1
    # set up your network and quantization style, the network arg. format should be: 'googlenet', 'resnet50', 'squeezenet', 'ssd' ...
    # the quantization style should be -- shift, simple or winograd
    style = sys.argv[1]
    net_name = sys.argv[2]
    q_path = 'channel_q/' + net_name
    model, layer_name, Features, feature_path = load_model(net_name)
    layer_num = len(layer_name)
    print('Start quantization: ')
    print('Total layers is: ',end='')
    print(layer_num)
    for m in range(layer_num):
        print('layer: ',end='')
        print(m)
        feature_shape = Features[m].shape
        dim = len(feature_shape)
        if dim == 1:
            with open(feature_path + '/' + layer_name[m] + '.bin',"rb") as bin_reader:
                for i in range(feature_shape[0]):
                        data = bin_reader.read(4)
                        data_float = struct.unpack("f",data)[0]
                        Features[m][i] = data_float
        if dim == 2:
            with open(feature_path + '/' + layer_name[m] + '.bin',"rb") as bin_reader:
                for i in range(feature_shape[0]):
                    for j in range(feature_shape[1]):
                        data = bin_reader.read(4)
                        data_float = struct.unpack("f",data)[0]
                        Features[m][i][j] = data_float
        if dim == 3:
            with open(feature_path + '/' + layer_name[m] + '.bin',"rb") as bin_reader:
                for i in range(feature_shape[0]):
                    for j in range(feature_shape[1]):
                        for k in range(feature_shape[2]):
                            data = bin_reader.read(4)
                            data_float = struct.unpack("f",data)[0]
                            Features[m][i][j][k] = data_float
        if dim == 4:
            with open(feature_path + '/' + layer_name[m] + '.bin',"rb") as bin_reader:
                for i in range(feature_shape[0]):
                    for j in range(feature_shape[1]):
                        for k in range(feature_shape[2]):
                            for l in range(feature_shape[3]):
                                data = bin_reader.read(4)
                                data_float = struct.unpack("f",data)[0]
                                Features[m][i][j][k][l] = data_float
        power = QuantizeChannel(style, Features[m])
        if not os.path.exists(q_path):
            os.makedirs(q_path)
        with open(q_path + '/' + layer_name[m] + '.txt','w') as data:
            for item in power:
                item = int(item)
                item = str(item)
                item = item + ' '
                data.write(item)

File: Runtime_Engine/cnn/host/src/quantization.cpp
/* Copyright 2019 Inspur Corporation. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0
    
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "quantization.h"

void read_file(FILE *fp, std::string file_name) {
  fp = fopen(file_name.c_str(), "rb");
  if (!fp) {
    ERROR("Can not open file %s\n", file_name.c_str());
  }
}

void Quantization(char *q, float *input, char* file_name) {
  int offset = 0;
  // input data
  FILE *fp;
  fp = fopen(file_name, "r");
  if (!fp) {
    ERROR("Can not open file %s\n", file_name);
  }
     
  INFO("file_name=%s\n", file_name);

  for (int layer = 0; layer < NUM_LAYER + 1; layer++) {
    int conv_layer = layer == 0 ? 0 : layer - 1;
    
    int channel = layer == 0 ? 3 : kOutputChannels[conv_layer];

    // For yandi network, in layer 24 yandi_q actually has 8 channel data,
    // the other 24 channel data come from layer 12 and layer 18 for channel concatenation.
    if (getNetwork() == kYandi) {
      if (layer == 24) {
        channel = 8;
      }
    }

    for (int c = 0; c < channel; c++) {
      int q_value = 0;
      if (kIpoolEnable[conv_layer]) {
        q[offset + c] = q[ kInputLayer[conv_layer] * MAX_OUT_CHANNEL + c ];
      } else {
        fscanf(fp, "%d", &q_value);
        q[offset + c] = -q_value;
        if (kBranchTail[conv_layer]) {
          q[(NUM_CONVOLUTIONS + 1 + kConcatLayer[conv_layer]) * MAX_OUT_CHANNEL + kNStart[conv_layer] + c] = -q_value; // 1 is for input data
        }
      }
    }
    offset += MAX_OUT_CHANNEL;
  }
  fclose(fp);

  // For yandi network, concate layer12,layer18 and layer24's Q datat o layer24.
  if (getNetwork() == kYandi) {
    for (int c = 0; c < 8; c++) {
      q[24 * MAX_OUT_CHANNEL + 24 + c] = q[24 * MAX_OUT_CHANNEL + c];
      q[24 * MAX_OUT_CHANNEL + 16 + c] = q[18 * MAX_OUT_CHANNEL + c];
    }
    for (int c = 0; c < 16; c++) {
      q[24 * MAX_OUT_CHANNEL  + c] = q[12 * MAX_OUT_CHANNEL + c];
    }
  }
}
